src/tools/concurrent_cache.h

```diff
@@ -95,6 +95,113 @@ class ConcurrentCache
   std::mutex lock_;
 };
 
+
+/**
+  WeakStore represent a thread safe store (map) of weak ptr.
+  It allows to store weak_ptr from shared_ptr and retrieve shared_ptr from
+  potential non expired weak_ptr.
+  It is not limited in size.
+*/
+template<typename Key, typename Value>
+class WeakStore {
+  private: // types
+    typedef std::weak_ptr<Value> WeakValue;
+
+  public:
+    explicit WeakStore() = default;
+
+    std::shared_ptr<Value> get(const Key& key)
+    {
+      std::lock_guard<std::mutex> l(m_lock);
+      auto it = m_weakMap.find(key);
+      if (it != m_weakMap.end()) {
+        auto shared = it->second.lock();
+        if (shared) {
+          return shared;
+        } else {
+          m_weakMap.erase(it);
+        }
+      }
+      throw std::runtime_error("No weak ptr");
+    }
+
+    void add(const Key& key, std::shared_ptr<Value> shared)
+    {
+      std::lock_guard<std::mutex> l(m_lock);
+      m_weakMap[key] = WeakValue(shared);
+    }
+
+
+  private: //data
+    std::map<Key, WeakValue> m_weakMap;
+    std::mutex m_lock;
+};
+
+template <typename Key, typename RawValue>
+class ConcurrentCache<Key, std::shared_ptr<RawValue>>
+{
+private: // types
+  typedef std::shared_ptr<RawValue> Value;
+  typedef std::shared_future<Value> ValuePlaceholder;
+  typedef lru_cache<Key, ValuePlaceholder> Impl;
+
+public: // types
+  explicit ConcurrentCache(size_t maxEntries)
+    : impl_(maxEntries)
+  {}
+
+  // Gets the entry corresponding to the given key. If the entry is not in the
+  // cache, it is obtained by calling f() (without any arguments) and the
+  // result is put into the cache.
+  //
+  // The cache as a whole is locked only for the duration of accessing
+  // the respective slot. If, in the case of the a cache miss, the generation
+  // of the missing element takes a long time, only attempts to access that
+  // element will block - the rest of the cache remains open to concurrent
+  // access.
+  template<class F>
+  Value getOrPut(const Key& key, F f)
+  {
+    std::promise<Value> valuePromise;
+    std::unique_lock<std::mutex> l(lock_);
+    const auto x = impl_.getOrPut(key, valuePromise.get_future().share());
+    l.unlock();
+    if ( x.miss() ) {
+      // Try to get back the shared_ptr from the weak_ptr first.
+      try {
+        valuePromise.set_value(m_weakStore.get(key));
+      } catch(const std::runtime_error& e) {
+        try {
+          const auto value = f();
+          valuePromise.set_value(value);
+          m_weakStore.add(key, value);
+        } catch (std::exception& e) {
+          drop(key);
+          throw;
+        }
+      }
+    }
+
+    return x.value().get();
+  }
```

Comment:
My feeling is that a similar result can be achieved through composition (without needing to specialize `ConcurrentCache` for `shared_ptr`).

1. `ConcurrentCache<Key, std::shared_ptr<Value>> cachedRefs;`
2. `WeakStore<Key, Value> allRefs;`
4. `cachedRefs.getOrPut(key, allRefs.wrap(key, f));`

where `WeakStore<Key, Value>::wrap(const Key& key, F f)` returns an augmented functor that first looks for `key` in `WeakStore`  and calls `f()` only if the key was not found.

Reply:
Indeed.

The advantage of specialization is that it is a facto replacement for all ConcurrentCache storing shared_ptr.
The composition means it is the responsibility of the caller to do the composition.

I don't see why we may have a cache system that would not reuse a existing value.
I even think we should backport this in libzim. It could be use for the cluster and dirent cache. If a item (which keep a shared_ptr to dirent) exists somewhere. We should reuse the dirent even if it is not "in" the cache.

Reply:
> I even think we should backport this in libzim. It could be use for the cluster and dirent cache. If a item (which keep a shared_ptr to dirent) exists somewhere. We should reuse the dirent even if it is not "in" the cache.

It definitely makes sense for the cluster cache, but I don't think that it is a good idea for the dirent cache, since the count of dirents can be quite large and in a (quite realistic) worst case scenario all of them can end up being "tracked" by the `WeakStore`.

On the other hand, the dirent cache is based on the lower level `lru_cache`. If this optimization for `std::shared_ptr` can be justified to be always applicable then it probably must be pushed down to `lru_cache`. Otherwise, it can still be implemented in (or on top of) `lru_cache` as an optional feature.